File: riveredge-frontend/src/apps/kuaiplm/services/project-proposal.ts
/**
 * 项目建议书 API（R-15 #68）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/project-proposals';

export type ProjectProposalStatus =
  | 'draft'
  | 'pending'
  | 'approved'
  | 'issued'
  | 'rejected';

export interface ProjectProposal {
  id: number;
  uuid: string;
  proposal_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  title: string;
  summary?: string | null;
  customer_name?: string | null;
  expected_date?: string | null;
  supplier_id?: number | null;
  supplier_code?: string | null;
  supplier_name?: string | null;
  supplier_contact?: string | null;
  supplier_remark?: string | null;
  status: ProjectProposalStatus;
  remarks?: string | null;
  submitted_at?: string | null;
  approved_at?: string | null;
  issued_at?: string | null;
  issued_by_name?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface ProjectProposalPayload {
  project_id: number;
  title: string;
  summary?: string | null;
  customer_name?: string | null;
  expected_date?: string | null;
  supplier_id?: number | null;
  supplier_code?: string | null;
  supplier_name?: string | null;
  supplier_contact?: string | null;
  supplier_remark?: string | null;
  remarks?: string | null;
}

function unwrapList(res: unknown): { items: ProjectProposal[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as ProjectProposal[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const projectProposalApi = {
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    project_id?: number;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => {
    const res = await api.get(`${BASE}/${id}`);
    return res as ProjectProposal;
  },
  create: async (data: ProjectProposalPayload) => {
    const res = await api.post(BASE, data);
    return res as ProjectProposal;
  },
  update: async (
    id: number,
    data: Partial<
      Pick<ProjectProposalPayload, 'title' | 'summary' | 'customer_name' | 'expected_date' | 'remarks'>
    >,
  ) => {
    const res = await api.put(`${BASE}/${id}`, data);
    return res as ProjectProposal;
  },
  fillSupplier: async (
    id: number,
    data: {
      supplier_id?: number | null;
      supplier_code?: string | null;
      supplier_name: string;
      supplier_contact?: string | null;
      supplier_remark?: string | null;
    },
  ) => {
    const res = await api.put(`${BASE}/${id}/supplier`, data);
    return res as ProjectProposal;
  },
  submit: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/submit`);
    return res as ProjectProposal;
  },
  approve: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/approve`);
    return res as ProjectProposal;
  },
  reject: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/reject`);
    return res as ProjectProposal;
  },
  issue: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/issue`);
    return res as ProjectProposal;
  },
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};

File: riveredge-backend/src/apps/kuaiplm/models/project_proposal.py
"""项目建议书（R-15 #68）

销售发起 → 采购填写供应商 → 审批 → 下发研发。
"""

from tortoise import fields

from core.models.base import BaseModel


class ProjectProposal(BaseModel):
    """项目建议书。"""

    class Meta:
        table = "apps_kuaiplm_project_proposals"
        table_description = "快研发 - 项目建议书"
        unique_together = [("tenant_id", "proposal_code")]
        indexes = [
            ("tenant_id", "project_id"),
            ("tenant_id", "status"),
            ("tenant_id", "supplier_id"),
            ("uuid",),
        ]

    id = fields.IntField(pk=True)
    proposal_code = fields.CharField(max_length=50, description="建议书单号")
    project_id = fields.IntField(description="研发项目ID")
    project_code = fields.CharField(max_length=50, description="项目代号快照")
    project_name = fields.CharField(max_length=200, description="项目名称快照")
    title = fields.CharField(max_length=200, description="标题")
    summary = fields.TextField(null=True, description="建议内容摘要")
    customer_name = fields.CharField(max_length=200, null=True, description="客户名称")
    expected_date = fields.DateField(null=True, description="期望推进日期")
    # 采购填写
    supplier_id = fields.IntField(null=True, description="供应商ID")
    supplier_code = fields.CharField(max_length=80, null=True, description="供应商编码快照")
    supplier_name = fields.CharField(max_length=200, null=True, description="供应商名称快照")
    supplier_contact = fields.CharField(max_length=200, null=True, description="供应商联系方式")
    supplier_remark = fields.TextField(null=True, description="采购备注")
    status = fields.CharField(
        max_length=20,
        default="draft",
        description="draft/pending/approved/issued/rejected",
    )
    remarks = fields.TextField(null=True)
    submitted_at = fields.DatetimeField(null=True)
    approved_at = fields.DatetimeField(null=True)
    issued_at = fields.DatetimeField(null=True)
    issued_by = fields.IntField(null=True)
    issued_by_name = fields.CharField(max_length=100, null=True)
    deleted_at = fields.DatetimeField(null=True)
